### ISSUE TITLE
`object.__setattr__`  triggering PLC2801 outside of dunder methods

### ISSUE THREAD

**Issue by my1e5** [astral-sh/ruff#11636]
Related to [false positive on PLC2801 "unnecessary dunder... for __setattr__" #9584](https://github.com/astral-sh/ruff/issues/9584)

`object.__setattr__` is the recommended way to set attributes on frozen dataclasses. This was addressed in #9584 but PLC2801 is triggering when `object.__setattr__` is in non-dunder method.

See - https://play.ruff.rs/1a9602da-78ce-44c9-bbfe-b17b147da04b
```py
from dataclasses import dataclass, field


@dataclass(frozen=True)
class Foo:
    a: int
    b: int
    c: int = field(init=False)

    def __post_init__(self):
        object.__setattr__(self, "c", self.a + self.b)

    def foo(self):
        object.__setattr__(self, "c", self.a - self.b)    # <-------

    def __str__(self):
        object.__setattr__(self, "c", self.a * self.b) 
```
```
14:9: PLC2801 Unnecessary dunder call to `__setattr__`. Mutate attribute directly or use setattr built-in function.
```

**Comment by AlexWaygood** [comment]
Thanks for the report!

I'm curious, what's the purpose of mutating attributes on a frozen dataclass from arbitrary non-dunder methods? Wouldn't the idiomatic thing here be to make it a regular (not frozen) dataclass?

**Comment by AlexWaygood** [comment]
Having said that, we may want to consider relaxing the rules here anyway. `object.__setattr__` can be useful in various other situations (though it's quite niche) where you want to make sure that you're avoiding triggering custom logic in descriptor or `__setattr__` methods
